src/optimizer/query_to_operator_transformer.cpp

```diff
@@ -245,11 +243,72 @@ void QueryToOperatorTransformer::Visit(parser::InsertStatement *op) {
     op->select->Accept(this);
     insert_expr->PushChild(output_expr_);
     output_expr_ = insert_expr;
+    return;
+  }
+  // column_objects represents the columns for the current table as defined in
+  // its schema
+  auto column_objects = target_table->GetColumnObjects();
+  // INSERT INTO table_name VALUES (val1, val2, ...), (val_a, val_b, ...), ...
+  if (op->columns.empty()) {
+    for (const auto &values : op->insert_values) {
+      if (values.size() > column_objects.size()) {
+        throw CatalogException(
+            "ERROR:  INSERT has more expressions than target columns");
+      } else if (values.size() < column_objects.size()) {
+        for (oid_t i = values.size(); i != column_objects.size(); ++i) {
+          // check whether null values are allowed in the rest of the columns
+          if (column_objects[i]->IsNotNull()) {
+            // TODO: Add check for default value's existence for the current column
+            throw CatalogException(StringUtil::Format("ERROR:  null value in column \"%s\" violates not-null constraint", column_objects[i]->GetColumnName().c_str()));
+          }
+        }
+      }
+    }
   } else {
-    auto insert_expr = std::make_shared<OperatorExpression>(
-        LogicalInsert::make(target_table, &op->columns, &op->insert_values));
-    output_expr_ = insert_expr;
+    // INSERT INTO table_name (col1, col2, ...) VALUES (val1, val2, ...), ...
+    auto num_columns = op->columns.size();
+    for (const auto &tuple : op->insert_values) {  // check size of each tuple
+      if (tuple.size() > num_columns) {
+        throw CatalogException(
+            "ERROR:  INSERT has more expressions than target columns");
+      } else if (tuple.size() < num_columns) {
+        throw CatalogException(
+            "ERROR:  INSERT has more target columns than expressions");
+      }
+    }
+
+    // vector below contains column ids of the columns mentioned in the insert statement
+    std::vector<oid_t> specified;
+    
+    for (const auto col : op->columns) {
+      auto found =
+          std::find_if(column_objects.begin(), column_objects.end(),
+                        [&col](const decltype(column_objects)::value_type &x) {
+                          return col == x.second->GetColumnName();
+                        });
+      if (found == column_objects.end()) {
+        throw CatalogException(StringUtil::Format("ERROR:  column \"%s\" of relation \"%s\" does not exist", col.c_str(), target_table->GetTableName().c_str()));
+      } else {
+        specified.push_back(found->first);
```

Comment:
I believe the way you are checking columns is not correct.
1. The columns specified do not have to be in schema column order. The user can specify an arbitrary order for the columns to be inserted.

2. Because of 1. the assumption that the specified vector can be initialized using push_back, is not valid

Reply:
@pervazea I think it's correct. I tried specifying the columns in a different order and the insert ran correctly.

Reply:
Look more carefully at the code, and consider two scenarios:
1. If the columns are out of order, is the check on line 303 still correct?
2. What if you have fewer columns specified than in the table columns? 

Maybe it will be easier to see if you add test cases to cover such cases. Be sure to run with the address sanitizer enabled if (when) you exercise these cases.

Reply:
@pervazea Thanks for pointing that out. The check at line 303 is still correct, however, the `index` increment isn't. I think I did keep this in mind while adding my code in the insert plan. However, when moving my code from insert plan to the transformer, the sort for `specified` vector got removed. Sorting the `specified` vector at line 295 will solve this problem.